tree/ntuple/v7/inc/ROOT/RNTupleDescriptor.hxx

```diff
@@ -455,6 +452,94 @@ public:
+
+// clang-format off
+/**
+\class ROOT::Experimental::RDanglingFieldDescriptor
+\ingroup NTuple
+\brief A helper class for piece-wise construction of an RFieldDescriptor
+
+Dangling field descriptors describe a single field in isolation. They are
+missing the necessary relationship information (parent field, any child fields)
+required to describe a real NTuple field.
+
+Dangling field descriptors can only become actual descriptors when added to an
+RNTupleDescriptorBuilder instance and then linked to other fields.
+*/
+// clang-format on
+class RDanglingFieldDescriptor {
+private:
+   RFieldDescriptor fField = RFieldDescriptor();
+public:
+   /// Make an empty dangling field descriptor.
+   RDanglingFieldDescriptor() = default;
+   /// Make a new RDanglingFieldDescriptor based off an existing descriptor.
+   /// Relationship information is lost during the conversion to a
+   /// dangling descriptor:
+   /// * Parent id is reset to an invalid id.
+   /// * Field children ids are forgotten.
+   ///
+   /// These properties must be set using RNTupleDescriptorBuilder::AddFieldLink().
+   explicit RDanglingFieldDescriptor(const RFieldDescriptor& fieldDesc);
+
+   /// Make a new RDanglingFieldDescriptor based off a live NTuple field.
+   static RDanglingFieldDescriptor FromField(const Detail::RFieldBase& field);
+
+   RDanglingFieldDescriptor& FieldId(DescriptorId_t fieldId) {
+      fField.fFieldId = fieldId;
+      return *this;
+   }
+   RDanglingFieldDescriptor& FieldVersion(const RNTupleVersion& fieldVersion) {
+      fField.fFieldVersion = fieldVersion;
+      return *this;
+   }
+   RDanglingFieldDescriptor& TypeVersion(const RNTupleVersion& typeVersion) {
+      fField.fTypeVersion = typeVersion;
+      return *this;
+   }
+   RDanglingFieldDescriptor& FieldName(const std::string& fieldName) {
+      fField.fFieldName = fieldName;
+      return *this;
+   }
+   RDanglingFieldDescriptor& FieldDescription(const std::string& fieldDescription) {
+      fField.fFieldDescription = fieldDescription;
+      return *this;
+   }
+   RDanglingFieldDescriptor& TypeName(const std::string& typeName) {
+      fField.fTypeName = typeName;
+      return *this;
+   }
+   RDanglingFieldDescriptor& NRepetitions(std::uint64_t nRepetitions) {
+      fField.fNRepetitions = nRepetitions;
+      return *this;
+   }
+   RDanglingFieldDescriptor& Structure(const ENTupleStructure& structure) {
+      fField.fStructure = structure;
+      return *this;
+   }
+   /// Attempt to make a field descriptor. This may fail if the dangling field
+   /// was not given enough information to make a proper descriptor.
+   RResult<RFieldDescriptor> GetDescriptor() const {
+      if (fField.GetId() == kInvalidDescriptorId) {
+         return R__FAIL("invalid field id");
+      }
+      if (fField.GetStructure() == ENTupleStructure::kInvalid) {
+         return R__FAIL("invalid field structure");
+      }
+      if (fField.GetFieldName() == "" && fField.GetId() != DescriptorId_t(0)) {
```

Comment:
The name check is a placeholder for this PR and I will change it to use `RFieldBase::EnsureValidFieldName` after #5967 lands.